pymtl3/stdlib/rtl/queues.py

```diff
@@ -156,3 +444,170 @@ def construct( s, MsgType, num_entries=2 ):
+
+class NormalQueue1EntryRTL( Component ):
+
+  def construct( s, EntryType ):
+
+    # Interface
+
+    s.enq = EnqIfcRTL( EntryType )
+    s.deq = DeqIfcRTL( EntryType )
+
+    # Components
+
+    s.entry = Wire( EntryType )
+    s.full  = Wire( Bits1 )
+
+    # Logic
+
+    @s.update_on_edge
+    def up_full():
+      if s.reset:
+        s.full = b1(0)
+      elif ~s.full and s.enq.en:
+        s.full = b1(1)
+      elif s.full and s.deq.en:
+        s.full = b1(0)
+      else:
+        s.full = s.full
+    
+    # TODO: figure out whether to use deepcopy here.
+    @s.update_on_edge
+    def up_entry():
+      s.entry = s.enq.msg if s.enq.en else s.entry
+
+    @s.update
+    def up_enq_rdy():
+      if ~s.reset:
+        s.enq.rdy = b1(1) if ~s.full else b1(0)
+      else:
+        s.enq.rdy = b1(0)
+
+    @s.update
+    def up_deq_rdy():
+      if ~s.reset:
+        s.deq.rdy = b1(1) if s.full else b1(0)
+      else:
+        s.deq.rdy = b1(0)
+
+    s.connect( s.entry, s.deq.msg )
+
+  def line_trace( s ):
+    return "{}({}){}".format( s.enq, s.full, s.deq )
+
+#-------------------------------------------------------------------------
+# PipeQueue1EntryRTL
+#-------------------------------------------------------------------------
+
+class PipeQueue1EntryRTL( Component ):
+
+  def construct( s, EntryType ):
+
+    # Interface
+
+    s.enq = EnqIfcRTL( EntryType )
+    s.deq = DeqIfcRTL( EntryType )
+
+    # Components
+
+    s.entry = Wire( EntryType )
+    s.full  = Wire( Bits1 )
+
+    # Logic
+
+    @s.update_on_edge
+    def up_full():
+      if s.reset:
+        s.full = b1(0)
+      elif ~s.full and s.enq.en and ~s.deq.en:
+        s.full = b1(1)
+      elif s.full and ~s.enq.en and s.deq.en:
+        s.full = b1(0)
+      else:
+        s.full = s.full
+    
+    # TODO: figure out whether to use deepcopy here.
+    @s.update_on_edge
+    def up_entry():
+      s.entry = s.enq.msg if s.enq.en else s.entry
+
+    @s.update
+    def up_enq_rdy():
+      if ~s.reset:
+        s.enq.rdy = b1(1) if ~s.full or s.deq.en else b1(0)
+      else:
+        s.enq.rdy = b1(0)
+
+    @s.update
+    def up_deq_rdy():
+      if ~s.reset:
+        s.deq.rdy = b1(1) if s.full else b1(0)
+      else:
+        s.deq.rdy = b1(0)
+
+    s.connect( s.entry, s.deq.msg )
+
+  def line_trace( s ):
+    return "{}({}){}".format( s.enq, s.full, s.deq )
+
+#-------------------------------------------------------------------------
+# BypassQueue1EntryRTL
+#-------------------------------------------------------------------------
+
+class BypassQueue1EntryRTL( Component ):
+
+  def construct( s, EntryType ):
+
+    # Interface
+
+    s.enq = EnqIfcRTL( EntryType )
+    s.deq = DeqIfcRTL( EntryType )
+
+    # Components
+
+    s.entry = Wire( EntryType )
+    s.full  = Wire( Bits1 )
+
+    # Logic
+
+    @s.update_on_edge
+    def up_full():
+      if s.reset:
+        s.full = b1(0)
+      elif ~s.full and s.enq.en and ~s.deq.en:
```

Comment:
the simplified expression is  ```s.full  = ~s.deq.en & (s.enq.en | s.full) ```

Reply:
well something like this is also fine .. I guess this will create more bridges for simulation though. Might as well just have simplified form in stdlib

Reply:
if you do k-map it's actually `s.full = ~s.deq.en & (s.enq.en | s.full) | s.enq.en & s.full` (or `s.full = s.enq.en & ~s.deq.en | s.enq.en & s.full | ~s.deq.en & s.full`).

Reply:
for single entry bypass queue your expression is correct because enq.en and deq.en when s.full is a don't care.